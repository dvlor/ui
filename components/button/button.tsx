import { emits, PropTypes } from '../_utils/vueExtend'
import { defineComponent, ExtractPropTypes, PropType } from 'vue'
import { installWrap } from '../_utils/types'
import { Prefix } from '../_utils/prefix'
import { ripple } from '../_utils/ripple'

const props = () => ({
  // 大小
  size: PropTypes.oneOf(['smll', 'default', 'large']),
  // 是否禁用
  disabled: PropTypes.bool,
  // 加载中
  loading: PropTypes.bool,
  // 图标
  icon: PropTypes.string,
  // 防抖时间 小于等于0不做处理
  debounceTime: PropTypes.number,
  // 形状
  shape: PropTypes.oneOf(['default', 'circle']),
  // 类型
  type: PropTypes.oneOf(['default', 'primary', 'dashed', 'link', 'text']),
  danger: PropTypes.bool,
  block: PropTypes.bool,
  ghost: PropTypes.bool,
  href: PropTypes.string,
  target: PropTypes.string,
  onClick: {
    type: Function as PropType<(e: MouseEvent) => void>
  }
})

export type ButtonProp = Partial<ExtractPropTypes<ReturnType<typeof props>>>

export const Button = installWrap(
  'button',
  defineComponent({
    components: { ripple },
    emits: emits(['click']),
    props: props(),
    slots: ['icon'],
    setup(props, { emit }) {
      const handleClick = (e: MouseEvent) => {
        // e.stopPropagation()
        if (props.disabled) {
          return
        }
        if (props.type == 'link' && props.href) {
          window.open(props.href, props.target || 'self')
        }
        emit('click')
      }

      let className = {
        [`${Prefix.classPrefix}button`]: true,
        [`${Prefix.classPrefix}button-danger`]: props.danger,
        [`${Prefix.classPrefix}button-${props.type}`]: !!props.type,
        [`${Prefix.classPrefix}button-ghost`]: !!props.ghost,
        [`${Prefix.classPrefix}button-disabled`]: !!props.disabled,
        [`${Prefix.classPrefix}button-block`]: !!props.block
      }

      return {
        handleClick,
        showRipple: props.type !== 'text' && props.type !== 'link' && !props.disabled,
        prop: {
          class: className
        }
      }
    },
    render() {
      return this.showRipple ? (
        <ripple>
          <div {...this.prop} onClick={this.handleClick}>
            {this.$slots.icon && this.$slots.icon()}
            {this.$slots.default()}
          </div>
        </ripple>
      ) : (
        <div {...this.prop} onClick={this.handleClick}>
          {this.$slots.icon && this.$slots.icon()}
          {this.$slots.default()}
        </div>
      )
    }
  })
)
