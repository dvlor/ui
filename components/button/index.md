---
category: Components
type: 通用
title: Button
subtitle: 按钮
cover: https://gw.alipayobjects.com/zos/alicdn/fNUKzY1sk/Button.svg
---

按钮用于开始一个即时操作。

## 何时使用

标记了一个（或封装一组）操作命令，响应用户点击行为，触发相应的业务逻辑。

## API

通过设置 Button 的属性来产生不同的按钮样式，推荐顺序为：`type` -> `shape` -> `size` -> `loading` -> `disabled`。

按钮的属性说明如下：

| 属性         | 说明                                                    | 类型               | 默认值    | 版本 |
| ------------ | ------------------------------------------------------- | ------------------ | --------- | ---- |
| size         | 按钮大小，可以设置为`smll \| default \| large`          | string             | `default` |      |
| disabled     | 是否禁用                                                | boolean            | `false`   |      |
| loading      | 加载中                                                  | boolean            | `false`   |      |
| icon         | 图标                                                    | string             | -         |      |
| debounceTime | 防抖时间 小于等于 0 不做处理                            | number             | -         |      |
| shape        | 形状 `default \| circle`                                | string             | `default` |      |
| type         | 按钮类型 `default \| primary \| dashed \| link \| text` | string             | `default` |      |
| loading      | 设置按钮载入状态                                        | boolean            | `false`   |      |
| danger       | 危险按钮                                                | boolean            | `false`   |      |
| block        | 自适应父元素宽度                                        | boolean            | `false`   |      |
| ghost        | 是否设置背景透明                                        | boolean            | `false`   |      |
| href         | 链接, type 为 `link` 时生效                             | string             | -         |      |
| target       | 同 a 标签 target 属性，type 为 `link` 时生效            | string             | -         |      |
| action       | 点击之后触发的行为,会自动显示 loading 状态              | (event) => Promise | -         |      |

### 事件

| 事件名称 | 说明             | 回调参数        | 版本 |
| -------- | ---------------- | --------------- | ---- |
| click    | 点击按钮时的回调 | (event) => void |      |

支持原生 button 的其他所有属性。
