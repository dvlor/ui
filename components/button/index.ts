export { Button } from './button'
export type { ButtonProp } from './button'

export { ButtonGroup } from './group'
export type { ButtonGroupProp } from './group'
