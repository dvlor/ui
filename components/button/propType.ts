import { PropTypes } from '../_utils/vueExtend'
import type { PropType, ExtractPropTypes } from 'vue'
import { tuple } from 'components/_utils/types'

const type = tuple('default', 'primary', 'dashed', 'link', 'text')
const size = tuple('small', 'default', 'large')
const shape = tuple('default', 'circle')

const buttonProp = () => ({
  // 类型
  type: PropTypes.oneOf(type).def('default'),
  // 大小
  size: PropTypes.oneOf(size).def('default'),
  // 是否禁用
  disabled: PropTypes.bool,
  // 加载中
  loading: PropTypes.bool,
  // 图标
  icon: PropTypes.string,
  // 防抖时间 小于等于0不做处理
  debounceTime: PropTypes.number,
  // 形状
  shape: PropTypes.oneOf(shape).def('default'),
  // 危险按钮
  danger: PropTypes.bool,
  // 块状按钮
  block: PropTypes.bool,
  // 背景透明按钮
  ghost: PropTypes.bool,
  // 超链接地址
  href: PropTypes.string,
  // 超链接打开目标
  target: PropTypes.string,
  // 点击动作，封装了loading进去
  action: {
    type: Function as PropType<(e: MouseEvent) => Promise<any>>
  }
})

export type ButtonProp = Partial<ExtractPropTypes<ReturnType<typeof buttonProp>>>

export { buttonProp }

const buttonGroupProp = () => ({
  // 大小
  size: PropTypes.oneOf(size).def('default'),
  onClick: {
    type: Function as PropType<(e: MouseEvent) => void>
  }
})

export type ButtonGroupProp = Partial<ExtractPropTypes<ReturnType<typeof buttonGroupProp>>>

export { buttonGroupProp }
