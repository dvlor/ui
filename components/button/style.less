@import '../style/theme/index.less';
@import '../style/mixins/index.less';

@btn-prefix-cls: ~'@{class-prefix}-button';

.@{btn-prefix-cls} {
  .transition();
  box-sizing: border-box;
  display: inline-block;
  padding: calc(@font-size-base - 6px) calc(@font-size-base - 4px);
  cursor: pointer;
  border-radius: 3px;
  color: @text-color;
  border: solid 1px @border-color-base;
  font-size: @font-size-base;
  background-color: @white;
  position: relative;

  &.@{btn-prefix-cls}-block{
    display: block;
    text-align: center;
    + .@{btn-prefix-cls}.@{btn-prefix-cls}-block{
      margin-left: unset;
      margin-top: calc(@font-size-base / 2);
    }
  }

  &:hover:not(.@{btn-prefix-cls}-disabled) {
    border-color: @primary-color;
    color: @primary-color;
  }

  &:active:not(.@{btn-prefix-cls}-disabled) {
    border-color: @primary-color-active;
    color: @primary-color-active;
  }

  + .@{btn-prefix-cls}{
    margin-left: calc(@font-size-base / 2);
  }

  &.@{btn-prefix-cls}-primary{
    background-color: @primary-color;
    color: @white;
    border: solid 1px @primary-color;

    &:hover:not(.@{btn-prefix-cls}-disabled) {
      background-color: @primary-color-hover;
      border-color: @primary-color-hover;
      color: @white;
    }

    &:active:not(.@{btn-prefix-cls}-disabled)  {
      background-color: @primary-color-active;
      border-color: @primary-color-active;
    }
    &.@{btn-prefix-cls}-ghost {
      color: @primary-color;
      &:hover:not(.@{btn-prefix-cls}-disabled) {
        border-color: @primary-color-hover;
        color: @primary-color-hover;
      }

      &:active:not(.@{btn-prefix-cls}-disabled)  {
        border-color: @primary-color-active;
        color: @primary-color-active;
      }
    }
  }
  &.@{btn-prefix-cls}-dashed{
    border-style: dashed;

    &:hover:not(.@{btn-prefix-cls}-disabled) {
      border-color: @primary-color;
      color: @primary-color;
    }

    &:active:not(.@{btn-prefix-cls}-disabled) {
      border-color: @primary-color-active;
    }

    &.@{btn-prefix-cls}-disabled {
      color: @disabled-color;
      background-color: @disabled-bg;
      border-color: @disabled-color;
    }
  }
  &.@{btn-prefix-cls}-link{
    color: @primary-color;
    border-width: 0;

     &:hover:not(.@{btn-prefix-cls}-disabled) {
      color: @primary-color-hover;
      text-decoration: underline;
    }

    &:active:not(.@{btn-prefix-cls}-disabled) {
      color: @primary-color-active;
    }

    &.@{btn-prefix-cls}-disabled{
      background-color: transparent;
    }
  }
  &.@{btn-prefix-cls}-text{
    color: @text-color;
    border-width: 0;


    &:active:not(.@{btn-prefix-cls}-disabled) {
      background-color: @item-hover-bg;
    }

    &.@{btn-prefix-cls}-disabled{
      background-color: transparent;
    }
  }
  &.@{btn-prefix-cls}-ghost{
    background-color: transparent !important;
  }
  
  &.@{btn-prefix-cls}-disabled{
    cursor: not-allowed;
    color: @disabled-color;
    background-color: @disabled-bg;
    border-color: @disabled-color;
  }
}