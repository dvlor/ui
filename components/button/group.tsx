import { emits } from '../_utils/vueExtend'
import { defineComponent } from 'vue'
import { withInstall } from '../_utils/types'
import { Prefix } from '../_utils/prefix'
import { ripple } from '../_utils/ripple'
import { buttonGroupProp } from './propType'

export type { ButtonGroupProp } from './propType'
export const ButtonGroup = withInstall(
  defineComponent({
    name: 'button-group',
    components: { ripple },
    emits: emits(['click']),
    props: buttonGroupProp(),
    setup() {
      let className = {
        [`${Prefix.classPrefix}button-group`]: true
      }

      return {
        prop: {
          class: className
        }
      }
    },
    render() {
      const child = this.$slots.default()
      return <div {...this.prop}>{child}</div>
    }
  })
)
