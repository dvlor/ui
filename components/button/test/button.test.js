import { Button } from '../button'
import { render, fireEvent } from '@testing-library/vue'

describe('components/button', () => {
  it('render', () => {
    expect(render(<Button>click</Button>).html()).toMatchSnapshot()
  })

  it('click event', async () => {
    let i = 0
    const handleClick = () => {
      i++
    }

    const { getByText } = render(<Button onClick={handleClick}>click me</Button>)
    await fireEvent.click(getByText('click me'))
    expect(i).toBe(1)
  })
})
