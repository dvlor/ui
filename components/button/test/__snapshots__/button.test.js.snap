// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/button render 1`] = `
"<div class=\\"u-button\\">
  <!---->click
</div>"
`;
